nixos/modules/services/logging/packetbeat.nix

```diff
@@ -0,0 +1,256 @@
+            };
+
+            fields = mkOption {
+              type = format.type;
+              description = ''
+                Extra fields that packetbeat should add to the top-level of the objects it
+                sends to its output.
+              '';
+            };
+
+            processors = mkOption {
+              type = format.type;
+              description = ''
+                Configuration of what enrichment and/or filtering processors packetbeat
+                should gather information from before sending it to the output.
+                See <link xlink:href='https://www.elastic.co/guide/en/beats/packetbeat/current/filtering-and-enhancing-data.html'/>
+                for the configuration options available.
+              '';
+            };
+          };
+        };
+        default = {};
+        description = ''
+          Any other configuration options you want to add.
+        '';
+        example = ''
+          output = {
+            elasticsearch = {
+              hosts = [ "localhost:9200" ];
+            };
+          };
+          setup = {
+            template = {
+              settings = {
+                index.number_of_shards =  1;
+              };
+            };
+            kibana = {
+              host =  "localhost:5601";
+            };
+          };
+          processors =  [
+            '''
+              if.contains.tags: forwarded
+              then:
+                - drop_fields:
+                  fields: [host]
+              else:
+                - add_host_metadata: ~
+            '''
+            "add_cloud_metadata: ~"
+            "add_docker_metadata: ~"
+          ];
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.packetbeat.settings = {
+      packetbeat = {
+        interfaces.device = mkDefault "any";
+        flows = {
+          timeout = mkDefault "30s";
+          period = mkDefault "10s";
+        };
+        protocols = mapAttrs (name: mkDefault) {
+          icmp = {
+            enabled = true;
+          };
+          amqp = {
+            ports = [ 5672 ];
+          };
+          cassandra = {
+            ports = [ 9042 ];
+          };
+          dhcpv4 = {
+            ports = [ 67 68 ];
+          };
+          dns = {
+            ports = [ 53 ];
+          };
+          http = {
+            ports = [ 80 8080 8000 5000 8002 ];
+          };
+          memcache = {
+            ports = [ 11211 ];
+          };
+          mysql = {
+            ports = [ 3306 3307 ];
+          };
+          pgsql = {
+            ports = [ 5432 ];
+          };
+          redis = {
+            ports = [ 6379 ];
+          };
+          thrift = {
+            ports = [ 9090 ];
+          };
+          mongodb = {
+            ports = [ 27017 ];
+          };
+          nfs = {
+            ports = [ 2049 ];
+          };
+          tls = {
+            ports = [ 443 993 995 5223 8443 8883 9243 ];
+          };
+        };
```

Comment:
@Infinisil As we discussed on IRC yesterday, apparently this "dictionary-style" (aka attrset of attrsets in Nix) is deprecated by packetbeat and packetbeat wants it to be a list of attrsets. So technically it should be easy to change this attrset of attrsets into a list of attrsets, but as you noted, that will make it immutable once set, so this mean that we somehow have to take the attrset of attrsets and convert it into a list of attrsets, to keep the possibility of the user to actually change this, while still conforming to the wishes of packetbeat.

Reply:
Do you have a link to where this deprecation is explained?

Reply:
I will see what I can find, so far, the only hint I've gotten was a warning in the log when running packetbeat with the generated config.

Reply:
The only reference I can find is this github issue: https://github.com/elastic/beats/pull/3518, the dictionary style should've been removed in 7.0.0 but for some reason it still works, but throws that error message. It was apparently deprecated already by 6.0.0. Apparently it was done so that you can define the same protocol analyzer several times, to get different settings, i.e. if you want one for regular HTTP and one for REST endpoints, like elasticsearch endpoint and so, which the dictionary style didn't allow because of merging.

Reply:
The actual deprecation notice is at the end of this: https://www.elastic.co/guide/en/beats/libbeat/master/release-notes-6.0.0-alpha1.html

Reply:
Might there be some way that we can ask the yaml generator to generate a list of objects instead of a dictionary from our attrset of attrsets? or some type of trickery with map?